Add admin censor message

fernandoabolafio opened the pull request:
closes #97 
This PR intends to accomplish:
- ~~Provide a receipt on ProposalSetStatus (server signature of the admin's signature)~~
- [x] Store an Admin's message on every proposal status change (required only when transiting from unvetted to censored)
- [x] Policy validation of admin's message 
- [x] Retrieve the last status change message within the proposal data

marcopeereboom commented on politeiawww/api/v1/v1.go:
```diff
@@ -567,6 +574,7 @@ type PolicyReply struct {
 	ProposalNameSupportedChars []string `json:"proposalnamesupportedchars"`
 	MaxCommentLength           uint     `json:"maxcommentlength"`
 	BackendPublicKey           string   `json:"backendpublickey"`
+	MinCensorMessageLength     uint     `json:"mincensormessagelength"`
```
Reason is omitempty so this makes no sense. Please remove.

marcopeereboom commented on politeiawww/api/v1/v1.go:
```diff
@@ -93,6 +93,9 @@ const (
 	// for the routes that return lists of proposals
 	ProposalListPageSize = 20
 
+	// PolicyMinCensorMessageLength is the min length of a censor message
```
Kill this. See below for more details.

marcopeereboom commented on politeiawww/api/v1/v1.go:
```diff
@@ -132,6 +135,7 @@ const (
 	ErrorStatusCannotVerifyPayment         ErrorStatusT = 35
 	ErrorStatusDuplicatePublicKey          ErrorStatusT = 36
 	ErrorStatusInvalidPropVoteStatus       ErrorStatusT = 37
+	ErrorMalformedCensorMessage            ErrorStatusT = 38
```
If you kill above then you won't need this either.

marcopeereboom commented on politeiawww/api/v1/v1.go:
```diff
@@ -502,8 +508,9 @@ type ProposalDetailsReply struct {
-	Signature      string      `json:"signature"` // Signature of Token+string(ProposalStatus)
+	Signature      string      `json:"signature"` // Signature of Token+string(ProposalStatus)+message
 	PublicKey      string      `json:"publickey"`
+	Message        string      `json:"message"` // Admin's message
```
Make this omitempty too

marcopeereboom commented on politeiawww/backend.go:
```diff
@@ -1585,17 +1586,26 @@ func (b *backend) ProcessNewProposal(np www.NewProposal, user *database.User) (*
 		return nil, err
 	}
 
+	// Validate censor message when the proposal is being censored
```
Kill this.

marcopeereboom commented on politeiawww/backend.go:
```diff
@@ -2407,6 +2418,7 @@ func (b *backend) ProcessPolicy(p www.Policy) *www.PolicyReply {
 		MaxProposalNameLength:      www.PolicyMaxProposalNameLength,
 		ProposalNameSupportedChars: www.PolicyProposalNameSupportedChars,
 		MaxCommentLength:           www.PolicyMaxCommentLength,
+		MinCensorMessageLength:     www.PolicyMinCensorMessageLength,
```
Kill this.

marcopeereboom commented on politeiawww/backend_proposal_test.go:
```diff
@@ -459,7 +460,8 @@ func TestCensoredProposal(t *testing.T) {
+	csrMessage := "unappropriated content"
+	censorProposal(b, npr.CensorshipRecord.Token, csrMessage, t, user, id)
 	pdr := getProposalDetails(b, npr.CensorshipRecord.Token, t)
 	verifyProposalDetails(np, pdr.Proposal, t)
```
Missing check?

fernandoabolafio replied:
This isn't possible right now because the message is only stored on d and the inventory isn't updating the record's metadata as reported here https://github.com/decred/politeia/issues/396. Would be better to add the check once the fix for the inventory gets merged in.

fernandoabolafio replied:
A verification of it isn't possible right now because the message is only stored on d and the inventory isn't properly updating the proposal metadata as reported here https://github.com/decred/politeia/issues/396.

fernandoabolafio commented:
@marcopeereboom thanks for the review. I addressed your concerns on my last commit. If it still not being as you want then I agree that the best it for you to write the diff.

fernandoabolafio commented:
closed on https://github.com/decred/politeia/pull/439
